### PR TITLE
Add ArduinoSpoof.h

### LINKED ISSUE #50: Spoof Arduino functions on desktop

### PR DESCRIPTION
**Pull request by blhough**
spoofs the common arduino functions so they can be compiled on desktop

closes #50

### CODE REVIEW

File: ArduinoSpoof.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#define VERBOSE_ARDUINO
+
+#ifdef VERBOSE_ARDUINO
+#define verbose_print(str) std::cout << str << std::endl;
+#else
+#define verbose_print(str)
+#endif // VERBOSE_ARDUINO
+
+void pinMode(uint8_t pin, uint8_t mode) {
+#ifdef VERBOSE_ARDUINO
```
Comment (ids1024) [inline ArduinoSpoof.h]: If `VERBOSE_ARDUINO` is not set, these functions result in `unused parameter` errors.

File: ArduinoSpoof.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+#define OUTPUT 1
+#define INPUT_PULLUP 2
+#define INPUT_PULLDOWN 3
+#define OUTPUT_OPENDRAIN 4
+#define INPUT_DISABLE 5
+
+#define VERBOSE_ARDUINO
+
+#ifdef VERBOSE_ARDUINO
+#define verbose_print(str) std::cout << str << std::endl;
```
Comment (ids1024) [inline ArduinoSpoof.h]: The macro isn't used

**Review by ids1024 — CHANGES_REQUESTED** [review]
When I try to build with this using `-include`, it results in "multiple definition" errors. Perhaps the implementations should be moved to a `.cpp` file?

**Comment by blhough** [comment]
Requested changes addressed. https://github.com/RoboticsClubAtUCDavis/micromouse/pull/51/commits/8fc87ab8c9693a205b19bebde0cd752bb21f2a9e

**Comment by ids1024** [comment]
You haven't addressed the "multiple definition" errors when building with `-include`.

**Comment by blhough** [comment]
Are you talking about when you build for the Teensy?

If so, add #define ARDUINO_BUILD to the main file.

**Comment by blhough** [comment]
I changed the #ifndef to something that becomes defined automatically when building on teensy vs having to manually define it in the main file.

**Comment by ids1024** [comment]
If this header is included in more than one .cpp file (I think it's best to include it everywhere with `-include` to match the behavior of the Arduino/Teensy standard library), it will fail to link, because the functions are defined multiple times.

**Comment by blhough** [comment]
I'm kinda confused by what you're saying. Is the pragma once not working? Or are you including the file before the Teensy files?

**Comment by ids1024** [comment]
Pragma once only means the header will be included once in each compilation unit (.cpp file). But each .o will have a separate definition of each function, causing errors when they are linked together.

**Comment by blhough** [comment]
Okay how about now?I have successful compiled this when including it in multiple files and on the Teensy.

Also renamed to Arduino.h because on the Teensy `#include<Arduino.h>` is required in files other than main.
When compiling for the Teensy, this file will not be in the directory checked by the Arduino compiler so instead it will include the normal version.

**Comment by ids1024** [comment]
Including the header in two random files (`Maze.cpp` and `Direction.cpp`) results in these errors: https://gist.github.com/6a91518ef15e48c8b40764a25d622b98

The solution to that would be to only declare the functions in the header, and define them in a `.cpp` header. Do you not get similar errors in Visual Studio? Perhaps it's because scons is building intermediate `.o` files before linking? Anyway, a seperate .cpp file is the correct way to do this.

(Also you're missing a `#include <stdio.h>`)

Edit: I get the same error compiling and linking in one step (without intermedate `.o` files)

**Comment by blhough** [comment]
Ohh okay, I made the functions static to prevent multiple definitions

**Comment by ids1024** [comment]
This results in "unused function" errors. Should I just disable that warning/error?

Also, `srandom` is an issue on Linux; libc provides a function with that name, which again prevents compilation. (There's also a `random` function, but that isn't an issue, since it takes different arguments).

I don't know how best to deal with that. Perhaps rename the function to `_arduino_srandom` (or similar) and then `#define srandom _arduino_srandom`. Although that's not ideal...

**Comment by blhough** [comment]
Okay I decided not to be lazy and moved the definitions to a cpp. Maybe that will fix the not used warning idk? If not it will need to be suppressed.

I also removed `srandom` because arduino also has `randomSeed` which does the same thing

**Review by ids1024 — APPROVED** [review]

**Pull request merged**